app/models/concerns/language_and_schedule_references.rb

```diff
@@ -7,5 +7,17 @@ module LanguageAndScheduleReferences
 
     has_many :schedules, as: :scheduleable, dependent: :destroy
     accepts_nested_attributes_for :schedules
+
+    before_validation :build_schedules
+    validate :validate_schedule_count
+
+    def validate_schedule_count
+      errors.add(:schedules, 'too many') if schedules.size > 21
+      errors.add(:schedules, 'not enough') if schedules.size < 21
```

Comment:
actually it will only be evaluate in 50% of the error cases. In any validation that passes it will be run as well... 😉